_posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown

```diff
@@ -0,0 +1,105 @@
+
+## Proxy Servers, Web Application Firewalling, and CDN Cost Reduction
+
+One of the original drivers of this project was to get away from the very expensive cost of our CDN/WAF provider.
+
+As you can see in our network diagram, we accomplished this by creating our own proxy servers that run a modified version of the OWASP WAF software. How we did this is probably worthy of its own blog post, but the basic recipe was this:
+
+* Build a proxy server that sits in front of our production application
+* Add the default OWASP rules
+* Using DNS changes, route a small percentage of our live traffic to the proxy servers
+* Note the false positives and use them to tune the OWASP rule set to match our application traffic
+* Rinse, lather, repeat until we are comfortable routing 100% of our traffic through the new proxy servers
+
+The CDN component was even more straightforward. With our proxy servers in place, we verified that we were setting correct [expiration headers](http://httpd.apache.org/docs/current/mod/mod_expires.html) on our assets and then enabled [mod_disk_cache](https://httpd.apache.org/docs/2.4/mod/mod_cache_disk.html) on our proxy servers. Once we verified that assets were being served from our proxy servers without calls to our origin servers, we enabled a [Rails asset host](http://api.rubyonrails.org/classes/ActionView/Helpers/AssetUrlHelper.html) configuration on our production application to send all asset requests to a [CloudFront Distribution](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/distribution-working-with.html) whose origin server was our proxy server pool. This took all asset traffic off our expensive CDN provider without directing it to our origin servers.
+
+The final savings calculation looked like this:
+
+* Old CDN/WAF provider: $16,000 per month
+* CloudFront distribution and OWASP-enabled proxy servers: $400 per month (including traffic costs)
+
+That's a **97.5%** savings we were delighted to return to the operating budget of the project for more important things like developing new features!
+
+In other blog posts we discuss in more detail the complexities of supporting [SSL certificates for our government customers's hostnames]({% post_url 2016-07-11-using-lets-encrypt-for-a-multi-tenant-web-application %}) and how we managed to comply with the [DNSSEC requirement]({% post_url 2016-07-09-making-dnssec-work-for-an-aws-hosted-saas %}) for government agencies. Those posts go into much more technical detail and are worth reading if you're interested in how we solve the security challenges of providing SaaS services for thousands of government agencies.
```

Comment:
s/solve/solved/? I admit that I changed verb tenses continuously throughout the SSL post, but "solved" sounds more correct to me here.